lib/app_profiler/server.rb

```diff
@@ -0,0 +1,292 @@
+# frozen_string_literal: true
+
+require "socket"
+require "rack"
+require "tempfile"
+
+# This module provides a means to start a golang-inspired profile server
+# it is implemented using stdlib and Rack to avoid additional dependencies
+
+module AppProfiler
+  module Server
+    HTTP_OK = 200
+    HTTP_BAD_REQUEST = 400
+    HTTP_NOT_FOUND = 404
+    HTTP_NOT_ALLOWED = 405
+    HTTP_CONFLICT = 409
+
+    TRANSPORT_UNIX = "unix"
+    TRANSPORT_TCP = "tcp"
+
+    DEFAULTS = {
+      enabled: false,
+      transport: TRANSPORT_UNIX,
+      cors: true,
+      cors_host: "*",
+      port: 0,
+      duration: 30,
+    }
+
+    mattr_accessor :enabled, default: DEFAULTS[:enabled]
+    mattr_accessor :transport, default: DEFAULTS[:transport]
+    mattr_accessor :cors, default: DEFAULTS[:cors]
+    mattr_accessor :cors_host, default: DEFAULTS[:cors_host]
+    mattr_accessor :port, default: DEFAULTS[:port]
+    mattr_accessor :duration, default: DEFAULTS[:duration]
+
+    class ProfileApplication
+      class InvalidProfileArgsError < StandardError; end
+
+      def initialize
+        @semaphore = Thread::Mutex.new
+        @profile_running = false
+      end
+
+      def call(env)
+        handle(Rack::Request.new(env)).finish
+      end
+
+      private
+
+      def handle(request)
+        response = Rack::Response.new
+        if request.request_method != "GET"
+          response.status = HTTP_NOT_ALLOWED
+          response.write("Only GET requests are supported")
+          return response
+        end
+        case request.path
+        when "/profile"
+          begin
+            stackprof_args, duration = validate_profile_params(request.params)
+          rescue => e
```

Comment:
This is a catch-all which could result in unexpected return messages. Maybe we should be specific about what error we expect to catch here?